torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,246 @@
+import itertools
+from typing_extensions import Protocol
+import warnings
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Modules that lazily initialize parameters, or "lazy modules",
+    derive the shapes of their parameters from the first input(s)
+    to their forward method. Until that first forward they contain
+    :class:torch.nn.UninitializedParameters that should not be accessed
+    or used, and afterward they contain regular :class:torch.nn.Parameters.
+    Lazy modules are convenient since they don't require computing some
+    module arguments, like the in_features argument of a typical :class:torch.nn.Linear.
```

Comment:
This needs some additional updates and a few formatting corrections because the formatting doesn't copy+paste easily from the Github comments. For example, ":class:torch.nn.Linear" -> ":class:\`torch.nn.Linear\`".

Reply:
sorry, my mistake :)